Docs for logging pip requirements and databricks runtime (#4428)
mlflow/mlflow: docs/source/models.rst

input_example
  reference to an artifact with :ref:`input example <input-example>`.

databricks_runtime

>     Databricks runtime version and type, if the model was trained in a Databricks notebook or job.

Reviewer: @andreakress do we need to mention anything about which version of mlflow client this was added in, or reflecting the current state is fine?

Reviewer 2: So, for the OS docs, this will appear with the new release, right (1.18.0?). If I look at earlier doc versions, it will not appear? If so, no need to add the version it was added in.

Reviewer: That's right - i forgot you can look at earlier versions of docs. That makes sense!

Additional Logged Files
^^^^^^^^^^^^^^^^^^^^^^^
For environment recreation, we automatically log ``conda.yaml`` and ``requirements.txt`` files whenever a model is logged. These files can then be used to reinstall dependencies using either ``conda`` or ``pip``.

conda.yaml
    When saving a model, MLflow provides the option to pass in a conda environment parameter that can contain dependencies used by the model. If no conda environment is provided, a default environment is created based on the flavor of the model. This conda environment is then saved in ``conda.yaml``.
requirements.txt
    The requirements file is created from the `pip portion <https://www.anaconda.com/blog/using-pip-in-a-conda-environment>`_ of the ``conda.yaml`` environment specification. Additional pip dependencies can be added to ``requirements.txt`` by including them as a pip dependency in a conda environment and logging the model with the environment. 

> The following shows an example of saving a model with a specified conda environment and the corresponding content of the generated ``conda.yaml`` and ``requirements.txt`` files.

Reviewer: maybe "specified" -> "manually specified" to make it more clear that you don't need to do this? (we do say that above but we can't expect someone to read that once, probably briefly, and have it register :-))

Reviewer 3: +1 to this yeah

input_example
  reference to an artifact with :ref:`input example <input-example>`.

databricks_runtime

>     Databricks runtime version and type, if the model was saved in a Databricks notebook or job.

Reviewer 3: ```suggestion
    Databricks runtime version and type, if the model was trained in a Databricks notebook or job.
```

Additional Logged Files
^^^^^^^^^^^^^^^^^^^^^^^

> For environment recreation, two additional files are automatically logged whenever a model is saved. These files can then be used to reinstall dependencies through either ``conda`` or ``pip``.

Reviewer 3: ```suggestion
For environment recreation, we automatically log `conda.yaml` and `requirements.txt` files whenever a model is logged. These files can then be used to reinstall dependencies using either ``conda`` or ``pip``.
```

conda.yaml
    When saving a model, MLflow provides the option to pass in a conda environment parameter that can contain dependencies used by the model. If no conda environment is provided, a default environment is created based on the flavor of the model. This conda environment is then saved in ``conda.yaml``.
requirements.txt

>     The requirements file is created from the pip dependencies inside of the conda environment. Additional pip dependencies can be added to ``requirements.txt`` by including them as a pip dependency in a conda environment and logging the model with the environment. 

Reviewer 3: The first sentence is a little unclear imo - maybe we could rephrase like "the pip portion of the conda.yaml environment specification" and link to the conda docs on how `conda.yaml` is constructed

The following shows an example of saving a model with a specified conda environment and the corresponding content of the generated ``conda.yaml`` and ``requirements.txt`` files.

.. code-block:: py

    conda_env = {
        'channels': ['defaults', 'conda-forge'],
        'dependencies': [
            'python=3.8.8',
            'pip'],
        'pip': [
            'mlflow',
            'scikit-learn==0.23.2',
            'cloudpickle==1.6.0'
        ],
        'name': 'mlflow-env'
    }

>     mlflow.sklearn.save_model(model, "my_model", conda_env=conda_env)

Reviewer 3: Can we use the more canonical `log_model` here?

The written ``conda.yaml`` file:

.. code-block:: text

    channels:

>       - defaults

Reviewer 3: Let's drop `defaults` here - I think we'll stop using this soon?

Reviewer 4: We've dropped the `defaults` channel in https://github.com/mlflow/mlflow/pull/4393 :)
